routers/repo/repo.go

```diff
@@ -458,80 +455,79 @@ func RedirectDownload(ctx *context.Context) {
-		archiveType = git.TARGZ
-	default:
-		log.Trace("Unknown format: %s", uri)
+	if aReq == nil {
 		ctx.Error(404)
 		return
 	}
-	refName = strings.TrimSuffix(uri, ext)
 
-	if !com.IsDir(archivePath) {
-		if err := os.MkdirAll(archivePath, os.ModePerm); err != nil {
-			ctx.ServerError("Download -> os.MkdirAll(archivePath)", err)
-			return
+	complete := aReq.IsComplete()
+	ctx.JSON(200, map[string]interface{}{
+		"archiving": !complete,
+		"complete":  complete,
+	})
+}
+
+// Download an archive of a repository
+func Download(ctx *context.Context) {
+	uri := ctx.Params("*")
+	aReq := archiver_service.DeriveRequestFrom(ctx, uri)
+
+	downloadName := ctx.Repo.Repository.Name + "-" + aReq.GetArchiveName()
+	if !aReq.IsComplete() {
+		aReq = archiver_service.ArchiveRepository(aReq)
+		archiver_service.LockQueue()
+		for !aReq.IsComplete() {
+			archiver_service.WaitForCompletion()
 		}
+		archiver_service.UnlockQueue()
 	}
 
-	// Get corresponding commit.
-	var (
-		commit *git.Commit
-		err    error
-	)
-	gitRepo := ctx.Repo.GitRepo
-	if gitRepo.IsBranchExist(refName) {
-		commit, err = gitRepo.GetBranchCommit(refName)
-		if err != nil {
-			ctx.ServerError("GetBranchCommit", err)
-			return
-		}
-	} else if gitRepo.IsTagExist(refName) {
-		commit, err = gitRepo.GetTagCommit(refName)
-		if err != nil {
-			ctx.ServerError("GetTagCommit", err)
-			return
-		}
-	} else if len(refName) >= 4 && len(refName) <= 40 {
-		commit, err = gitRepo.GetCommit(refName)
-		if err != nil {
-			ctx.NotFound("GetCommit", nil)
-			return
-		}
+	if aReq.IsComplete() {
+		ctx.ServeFile(aReq.GetArchivePath(), downloadName)
 	} else {
-		ctx.NotFound("Download", nil)
+		ctx.Error(404)
+	}
+}
+
+// InitiateDownload will enqueue an archival request, as needed.  It may submit
+// a request that's already in-progress, but the archiver service will just
+// kind of drop it on the floor if this is the case.
+func InitiateDownload(ctx *context.Context) {
+	uri := ctx.Params("*")
+	aReq := archiver_service.DeriveRequestFrom(ctx, uri)
+
+	if aReq == nil {
 		return
 	}
 
-	archivePath = path.Join(archivePath, base.ShortSha(commit.ID.String())+ext)
-	if !com.IsFile(archivePath) {
-		if err := commit.CreateArchive(archivePath, git.CreateArchiveOpts{
-			Format: archiveType,
-			Prefix: setting.Repository.PrefixArchiveFiles,
-		}); err != nil {
-			ctx.ServerError("Download -> CreateArchive "+archivePath, err)
-			return
+	complete := aReq.IsComplete()
+	if !complete {
+		aReq = archiver_service.ArchiveRepository(aReq)
+		// As with the standard Download, we'll wait up to two seconds for the request
+		// to be completed.  The difference is that we'll never download the file from a POST
+		// request, only indicate the current status.  If we did manage to complete the request
+		// in this timeframe, the download will proceed with no further overhead.
+		timeout := time.Now().Add(2 * time.Second)
+		for {
+			if aReq.IsComplete() || time.Now().After(timeout) {
+				break
+			}
```

Comment:
This spin loop is taxing for the CPU and a probable door for a DOS attack. The "golang way" should be to use a `chan`, but we could make do with a thread sleep:
```
time.Sleep(250 * time.Millisecond)
```

Reply:
To be honest, what we really want here is a `sync.Cond.TimedWait` -- perhaps something like this: https://github.com/golang/go/issues/9578#issuecomment-69700942... this would entail the extra goroutine to Cond.Wait() and close the channel if it returns, but is that a problem?